thesis/03-implementation.tex

```diff
@@ -83,23 +83,491 @@ \subsubsection{Backend}
+
+\medskip
+Furthermore, no changes can be pushed directly to the \texttt{master} branch. Every commit needs to be reviewed by at least one other developer before being merged. This policy is enforced using \tech{Github} settings.
+
+\medskip
+Each system component has a dedicated pipeline, tailored to specific requirements. 
+
+\medskip
+Continuous delivery and deployment reduce the risk of deploying an application that has not passed all the required checks in a pipeline.
+
+\medskip
+All of the dependencies are bound to a specific minor version. Doing so ensures that in case one of the vendors implements a breaking change our application and pipelines do not break.
+
+\subsubsection{Frontend}
+
+The frontend code is linted on every commit using \tech{Husky}, \tech{ESLint}, and \tech{Prettier}. It is not possible to push an incorrectly formatted code to the repository.
+Uniformly formatted code improves the clarity, allowing a code reviewer to focus on business logic and potential bugs instead of code-formatting.
+
+\medskip
+Most of the business logic on the frontend is tested. The test runner of choice is \tech{Jest}.
+
+\medskip
+
+To aid the code review process, every pull request with changes related to the frontend component is deployed as a \tech{preview application} on \tech{Netlify}. A \tech{preview application} is an isolated environment with an application built using the code submitted along with a pull request.
+
+\medskip
+
+Running the whole test suite is as simple as executing \texttt{yarn test}.
+
+\begin{figure}[H]
+  \caption{Workflow - frontend component. Every commit is tested and built.}
+  \centering
+    \includegraphics[width=\textwidth]{assets/3-frontend-pipeline.png}
+\end{figure}
+
+\subsubsection{Solver}
+
+The Solver is a library written in python and tested using the \tech{unittest} module. Most of the business logic is covered. 
+
+The line test coverage is 77\%.
+
+\begin{figure}[H]
+  \caption{Workflow - solver component. Every change is tested.}
+  \centering
+    \includegraphics[width=0.8\textwidth]{assets/3-solver-pipeline.png}
+\end{figure}
+
+
+\subsubsection{Community}
+
+The community component is written in \tech{Python} using the \tech{Django} framework and therefore uses the \tech{unittest} module that is built-in to the Python standard library. All of the public-facing endpoints are fully covered with integration tests verifying that authentication, authorization, and request body handling are properly implemented. The line test coverage is 96\%.
+
+\medskip
+Code style abides by the \tech{PEP8} standard.
+
+\medskip
+To ensure build reproducibility, the community component is enclosed in a Docker image, which is automatically deployed to a production environment. 
+
+\begin{figure}[H]
+  \caption{Pipeline - Community backend component.}
+  \centering
+    \includegraphics[width=\textwidth]{assets/3-community-pipeline.png}
+\end{figure}
+
+\subsection{Deployment}
```

Comment:
First, there is some introduction about how the app is packaged. Ok.
Then, there is a local config file, but we are talking about the production? Hmm, strange. Maybe we should show 2 configurations - local and exemplar production. That way we could highlight the point that it is environment agnostic.

Then it says "Origuide backend can be deployed". And what about frontend?

In general, I'd structurize it a bit more. First, tell me about deploying this application in general. How it is packaged, how the deployment is structured, how the config files look like, how to deploy each component.
Then, tell me about the production environment and how awesome it is and why.

Reply:
Added note about deploying the frontend.

I wouldn't focus on the production environment. It's environment agnostic and cloud provider agnostic, therefore it doesn't make much sense in terms of the whole work.